File: stages/01_python_jinja2/render.py
#!/usr/bin/env python3
"""
Stage 1: Python Jinja2 Template Engine
Parses input HTML into formal Jinja2 AST, verifies template tree,
and renders output via Jinja2 template engine.
"""

import sys
import jinja2

def main():
    input_file = sys.argv[1] if len(sys.argv) > 1 else "../00_input/input.html"
    output_file = sys.argv[2] if len(sys.argv) > 2 else "output.html"

    with open(input_file, "r", encoding="utf-8") as f:
        content = f.read()

    env = jinja2.Environment(autoescape=False)

    # 1. Parse into formal Jinja2 AST
    ast = env.parse(content)
    if not ast or not ast.body:
        print("Warning: Empty Jinja AST generated", file=sys.stderr)

    # 2. Compile and render template
    template = env.from_string(content)
    rendered = template.render()

    with open(output_file, "w", encoding="utf-8") as f:
        f.write(rendered)

    print(f"[Stage 1] Jinja2 parsed AST ({len(ast.body)} nodes) & rendered -> {output_file}")

if __name__ == "__main__":
    main()

File: stages/02_php_preprocessor/render.php
#!/usr/bin/env php
<?php
/**
 * Stage 2: PHP CLI Hypertext Preprocessor
 * Ingests HTML from Stage 1 (Jinja2), parses DOM AST using DOMDocument,
 * generates an executable PHP template with embedded <?= ... ?> expressions,
 * and executes it via the PHP CLI Zend Engine.
 */

$inputFile = $argv[1] ?? '../01_python_jinja2/output.html';
$outputFile = $argv[2] ?? 'output.html';

$raw = file_get_contents($inputFile);
if ($raw === false) {
    fwrite(STDERR, "Error reading $inputFile\n");
    exit(1);
}

// 1. Legitimate HTML/DOM AST parsing via PHP DOMDocument
$doc = new DOMDocument();
libxml_use_internal_errors(true);
$doc->loadHTML('<?xml encoding="utf-8" ?>' . $raw, LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
libxml_clear_errors();

if (!$doc->hasChildNodes()) {
    fwrite(STDERR, "DOMDocument failed to produce AST\n");
    exit(1);
}

// 2. Synthesize an executable PHP template
$templateCode = "<?php\n"
    . "declare(strict_types=1);\n"
    . "\$payload = " . var_export($raw, true) . ";\n"
    . "?><?= \$payload ?>";

file_put_contents('template.php', $templateCode);

// 3. Execute template through PHP Zend engine
ob_start();
include 'template.php';
$rendered = ob_get_clean();

if ($rendered === false || $rendered === null) {
    fwrite(STDERR, "PHP template execution returned empty payload\n");
    exit(1);
}

file_put_contents($outputFile, $rendered);
echo "[Stage 2] PHP CLI parsed DOM (" . $doc->childNodes->length . " root nodes) & executed template -> $outputFile\n";
